Refactor `healthcheck` for Kubernetes extensibility with `HealthChecker` interface

rana opened the pull request:
This is part of Kubernetes health check integration.

The main intent of refactoring is to provide health check extensibility for Kubernetes while supporting the existing DB health checks. A `HealthChecker` interface is added to support the different health check approaches of DBs and Kubernetes. Existing DB TCP health check logic is moved to a new `TargetDialer` struct.

In this PR:
- Added `HealthChecker` interface with two functions:
    - `CheckHealth(ctx context.Context) ([]string, error)`
    - `GetProtocol() types.TargetHealthProtocol`
- Added `TargetDialer` struct which encapsulates existing TCP health check logic
- Changed `Target` struct to use the `HealthChecker` interface
- Changed `worker.checkHealth` to call the new `CheckHealth` function
- Removed a `protocol` field from `healthCheckConfig`
- Added `TargetHealthProtocolHTTP` for use with Kubernetes health checks
- Moved and renamed test `Test_dialEndpoints` to `TestTargetDialer_dialEndpoints`
- Added files `net.go` and `net_test.go` for `TargetDialer`

Relates to:
- #58413

Tener commented on lib/healthcheck/target.go:
```diff
@@ -20,51 +20,29 @@ package healthcheck
+	// CheckHealth checks the health of the target resource.
+	CheckHealth(ctx context.Context) error
+
+	// --- test methods below ---
+
+	// OnHealthCheck is called after each health check.
+	OnHealthCheck(lastResultErr error)
+	// OnConfigUpdate is called after each config update.
+	OnConfigUpdate()
+	// OnClose is called after the target's worker closes.
+	OnClose()
```
Having test methods in the interface is unusual. I think we may not need those?

Tener commented on lib/healthcheck/worker.go:
```diff
@@ -201,9 +193,7 @@ func (w *worker) run() {
-		if w.target.onClose != nil {
-			w.target.onClose()
-		}
+		w.target.OnClose()
```
Being the call site of the test callbacks and paired to targets 1-1, `worker` seems like a more natural choice for placing these callbacks in. That would change the call to something like this:

```suggestion
		w.onClose()
```

We would need to change `manager` struct to include testing fields, but that seems reasonable seeing as these callbacks are used in `manager_test.go` of all places.

rana replied:
For `w.onClose()`, some design adjustment for the `AddTarget` function to return a `worker` would be needed; or, some similar mechanism. `onClose` and would be set in the manager test. That also doesn't seem ideal. In which case, if we get away from the interface, restoring the Target as struct might be appropriate.

Here are are a couple options I see.

Option A:
Restore the existing `Target` to a struct, which has the test fields for use in the manager test. This would address concerns about interfaces with test methods. This is how the code is prior to this PR. 
```go
type Target struct {
	GetResource func() types.ResourceWithLabels
	ResolverFn EndpointsResolverFunc

	// -- test fields below --

	dialFn dialFunc
	onHealthCheck func(lastResultErr error)
	onConfigUpdate func()
	onClose func()
}

func (w *worker) run() {
    // ...
    w.checkHealth(w.closeContext)
    if w.target.onHealthCheck != nil {
        w.target.onHealthCheck(w.lastResultErr)
    }
}
```

Option B:
Another option can be to use a separate `Observable` interface which would be used in testing only. 
```go
// Observable is an optional interface for targets that need observation
type Observable interface {
    OnHealthCheck(lastResultErr error)
    OnConfigUpdate()
    OnClose()
}

func (w *worker) run() {
    // ...
    w.checkHealth(w.closeContext)
    if obs, ok := w.target.(Observable); ok {
        obs.OnHealthCheck(w.lastResultErr)
    }
}
```

In either case, some `if` check is present to support the test in `manager` code, making the approaches roughly equivalent in my mind. Some accommodation for test support is needed somewhere. I do understand the concern about exposing testing methods in the interface not being ideal as well. 

Why use an interface at all? The motivation is that DB target and Kube target have different scenarios which can be handled with the interface.

Currently proposed for the database in this PR:
```go
type TargetDialer struct {
	Resource func() types.ResourceWithLabels
	Resolver EndpointsResolverFunc
	Protocol types.TargetHealthProtocol
	lastResolvedEndpoints []string
	dial dialFunc

	// -- test fields below --

	onHealthCheck func(lastResultErr error)
	onConfigUpdate func()
	onClose func()
}

// startHealthCheck starts health checks for the database.
func (s *Server) startHealthCheck(ctx context.Context, db types.Database) error {
	resolver, err := s.getEndpointsResolver(ctx, db)
	if err != nil {
		return trace.Wrap(err)
	}
	targetDialer := &healthcheck.TargetDialer{
		Resource: func() types.ResourceWithLabels {
			s.mu.RLock()
			defer s.mu.RUnlock()
			return s.copyDatabaseWithUpdatedLabelsLocked(db)
		},
		Resolver: resolver,
	}
	err = s.cfg.healthCheckManager.AddTarget(targetDialer)
	return trace.Wrap(err)
}
```

And for Kube, the existing `kubeDetails` is the target, and passed to `AddTarget`.
```go
// startHealthCheck starts checking health for a Kubernetes cluster.
func (t *TLSServer) startHealthCheck(kube types.KubeCluster) error {
	kubeDetails, err := t.fwd.findKubeDetailsByClusterName(kube.GetName())
	if err != nil {
		return trace.Wrap(err)
	}
	err = t.healthCheckManager.AddTarget(kubeDetails)
	return trace.Wrap(err)
}
```

At this point I'm leaning towards using the `Observable` interface, or switching back to the Target struct.

@GavinFrazar What's your perspective?

Joerger commented on api/types/target_health.go:
```diff
@@ -25,8 +25,10 @@ import (
 const (
-	// TargetHealthProtocolTCP is a target health check protocol.
-	TargetHealthProtocolTCP TargetHealthProtocol = "TCP"
+	// TargetHealthProtocolTCP is the TCP target health check protocol.
+	TargetHealthProtocolTCP TargetHealthProtocol = "tcp"
```
Could changing this from uppercase cause any backwards compatibility issues? Is there any cross-service communication of this protocol that could have non-equal versions?

GavinFrazar reviewed (approved):
just a couple nitpicks but LGTM overall

Commit:
Refactor `healthcheck` for Kubernetes extensibility with `HealthChecker` interface

The main intent of refactoring is to provide health check extensibility for Kubernetes while supporting the existing DB health checks. A `HealthChecker` interface is added to support the different health check approaches of DBs and Kubernetes. Existing DB TCP health check logic is moved to a new `TargetDialer` struct.

Changes:
- Added `HealthChecker` interface with two functions:
    - `CheckHealth(ctx context.Context) ([]string, error)`
    - `GetProtocol() types.TargetHealthProtocol`
- Added `TargetDialer` struct which encapsulates existing TCP health check logic
- Changed `Target` struct to use the `HealthChecker` interface
- Changed `worker.checkHealth` to call the new `CheckHealth` function
- Removed a `protocol` field from `healthCheckConfig`
- Added `TargetHealthProtocolHTTP` for use with Kubernetes health checks
- Moved and renamed test `Test_dialEndpoints` to `TestTargetDialer_dialEndpoints`
- Added files `net.go` and `net_test.go` for `TargetDialer`

Part of #58413
